### ISSUE TITLE
Screen Reader announces terminal output over and over

### ISSUE THREAD

**Issue by MarcoZehe** [microsoft/vscode#95441]
- VSCode Version: 1.45 insider
- OS Version: Windows 10 19H2.

Steps to Reproduce:

1. Turn on a screen reader like NVDA.
2. In VS Code, Open a terminal. Listen as screen reader reads the initial PowerShell output, or Bash, or whichever terminal is being used.
3. Issue a command. Listen as the screen reader starts reading everything from the top instead of just reading the newly added text.

Does this issue occur when all extensions are disabled?: Yes

CC @isidorn

**Comment by isidorn** [comment]
I actually tried this out and could not reproduce both with VoiceOver and NVDA.
For me only the last output is being read. 
@MarcoZehe do you have some settings set which might influecne this. Can you also repro with some other screen reader?

Also assigning to @Tyriar so he tries to reproduce

**Comment by MarcoZehe** [comment]
Hm, OK, there seem to be two issues here. One, where I indeed issue a few commands and the output is spoken, I hear a lot of information twice. Once as the output, and once as the same output plus the reappearing prompt for input.

The second is that one parent, an item with classes "terminal xterm focus", contains a lot of output in its accessible name that is much older than what is currently being displayed in the grandchild list of outputs. That gets spoken whenever focus transitions into the terminal pane, for example via the F6 key. I was confused by these two sources of output, which at one point seemed to intermix probably due to me being too fast in switching panes and issuing commands.

But the double-speaking of output I can definitely reproduce reliably.

**Comment by jvesouza** [comment]
I also cannot reproduce using the orca.
I tested using the following command: cat / etc / passwd
Except for a message that I can't quite understand, everything seems to have worked well.
The message I didn't quite understand looks like this:
to much output to announce, navigate to rows manually to read.

**Comment by Tyriar** [comment]
> One, where I indeed issue a few commands and the output is spoken, I hear a lot of information twice.

@MarcoZehe do you have the `terminal.integrated.windowsEnableConpty` set to something?

> The second is that one parent, an item with classes "terminal xterm focus", contains a lot of output in its accessible name that is much older than what is currently being displayed in the grandchild list of outputs.

Do you mean the accessible list does not accurately reflect what's on screen?

> That gets spoken whenever focus transitions into the terminal pane, for example via the F6 key. I was confused by these two sources of output, which at one point seemed to intermix probably due to me being too fast in switching panes and issuing commands.

Does the entire terminal viewport read when you focus the terminal? The intent is just the prompt line is read.

**Comment by MarcoZehe** [comment]
> @MarcoZehe do you have the `terminal.integrated.windowsEnableConpty` set to something?

It is set to True/the checkbox is checked. I didn't make any change in this area, and I am using a new enough version of Windows 10.

> > The second is that one parent, an item with classes "terminal xterm focus", contains a lot of output in its accessible name that is much older than what is currently being displayed in the grandchild list of outputs.
> 
> Do you mean the accessible list does not accurately reflect what's on screen?

I think the *list* does, it's the grandparent whose accessible name reflects other stuff.

> Does the entire terminal viewport read when you focus the terminal? The intent is just the prompt line is read.

Not always. But sometimes, and sometimes it is read twice. Or the new text that comes in after I issue a command is read twice. Without being able to visually verify what is happening, this is hard to tell. I am totally blind, so cannot give you any visual correlation.

**Comment by Tyriar** [comment]
> It is set to True/the checkbox is checked. I didn't make any change in this area, and I am using a new enough version of Windows 10.

Ok, just checking. That means if you're on Windows 10 1909 that you're on the new system. For this it may just be a problem with how the terminal works on Windows as we announce whatever is coming from the backend that tries to emulate unix ptys, and sometimes it reprints the screen on Windows 10 for backwards compatibility reasons.

> I think the list does, it's the grandparent whose accessible name reflects other stuff.

Again is it the live region causing problems? The terminal announces to screen readers using alerts because not all data is rendered to the accessibility tree, consider when running the `tree` command for example. That's an enormous amount of data and we simply cannot add all of that to the DOM.

https://github.com/xtermjs/xterm.js/blob/c1e618fc3b4b8ace77289be3f1cfe0fb9c6b9366/src/AccessibilityManager.ts#L291-L297

@isidorn thoughts on this? Whenever anything is changed in the terminal, the accessibility tree is reconstructed (used for navigation mode) and any  _new_ characters are read via the alert by appending it to the alert's contents. Note also that this data comes in asynchonously in chunks, so we need to be able to append the data and for readers to announce it correctly if we only have "hel" initially and then add "lo" later for "hello".

**Comment by isidorn** [comment]
@Tyriar I think it makes sense to read new content that is coming in via an `alert`, because the focus did not change, all the time you are in the input.
Might it be due to the async some duplication happens?

Maybe we compare to what some other terminals are doing?
@marcozehe might have some details how mozilla does this for their dev tools terminal

**Comment by MarcoZehe** [comment]
We actually do it via an aria-live="polite" on the output container, in this case the div with role "list" that contains the output. And maybe experiment with the aria-atomic attribute here so stuff doesn't get read in a weird way.

**Comment by isidorn** [comment]
@MarcoZehe all good ideas, thanks!
@Tyriar can we incorporate maybe some of this? So at least change from `assertive` to `polite`. aria-atomic might aslo be interesting. More details here https://w3c.github.io/aria/#aria-atomic

**Comment by MarcoZehe** [comment]
I just again encountered this last night. The action I was taking was:

1. Open the VS Code checkout folder.
2. Open a terminal.
3. Enter `git pull upstream master`.
   * Result: That spoke OK.
4. Typed `yarn` to update dependencies.
   * Result: Output was repeated over and over, always starting with „Yarn install v1.19.2“...

So it seems to be a bit dependent on what commands are being issued. The yarn command in particular seems to be repeating text a lot.

Requesting that this be reopened.

**Comment by Tyriar** [comment]
So for yarn in particular, the way it works is by reprinting several lines constantly to indicate progress.

I know this is an awful experience and even proposed a way to solve the issue generally at terminal-wg by introducing something similar to aria label in html, it didn't get much traction though and would need widespread adoption to be useful https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/18

**Comment by Tyriar** [comment]
I don't think there's anything we can do here to action this. On Windows conpty sends us the text to print, and we announce it as it comes in. Under certain circumstances conpty may reprint the screen which would result in reading everything, but we can't tell the difference between that happening because of conpty or the backing process. The good news is that conpty has been steadily improving these edge cases over time so you should expect a generally better experience as you get updates to Windows.
